plugin-mainmenu/xdgcachedmenu.cpp

```diff
@@ -95,15 +96,32 @@ void XdgCachedMenu::addMenuItems(QMenu* menu, MenuCacheDir* dir)
     }
     else
     {
+      bool appVisible = type == MENU_CACHE_TYPE_APP
+          && menu_cache_app_get_is_visible(MENU_CACHE_APP(item),
+                                           menu_cache_desktop_env_flag_);
+      bool dirVisible = type == MENU_CACHE_TYPE_DIR
+          && menu_cache_dir_is_visible(MENU_CACHE_DIR(item));
+
+      if(!appVisible && !dirVisible)
+        continue;
+
       XdgCachedMenuAction* action = new XdgCachedMenuAction(item, menu);
       menu->addAction(action);
+
       if(type == MENU_CACHE_TYPE_APP)
         connect(action, SIGNAL(triggered(bool)), SLOT(onItemTrigerred()));
       else if(type == MENU_CACHE_TYPE_DIR)
       {
         XdgCachedMenu* submenu = new XdgCachedMenu(menu);
         action->setMenu(submenu);
-        addMenuItems(submenu, (MenuCacheDir*)item);
+        addMenuItems(submenu, MENU_CACHE_DIR(item));
+
+        if(submenu->actions().count() == 0)
```

Comment:
>> Agree completely. It's what menu_cache_dir_is_visible() is for so this part of code is never needed and not correct.

>Are you sure? Reading the code I think the function only checks if the entry is marked as NODISPLAY.

Yes, I am sure. That attibute is enforced by menu-cache-gen for empty
menu directory in case if show_empty isn't set in layout for that menu,
and empty menu directory even not included into cache file unless the
'+hidden' variant was requested (so hidden/nodisplay items are included
as well but marked as such).